HelpSource/Classes/CondVar.schelp

```diff
@@ -0,0 +1,414 @@
+point in the future. Signalling means indicating to other threads that the condition is now true, and that they should
+continue executing.
+
+When waiting on a CondVar, you may optionally pass a predicate (an object, usually a Function, that returns a
+Boolean when evaluated with code::value::) as an argument. This leads to two general ways of using CondVar:
+
+list::
+## Manually checking your condition
+
+	numberedlist::
+	## First, check the condition in case it is already true.
+	## If the condition is false, call wait or waitFor without a predicate. This suspends execution of the current thread.
+	## When the CondVar is signalled or a timeout expires, the thread continues execution. You should then check the
+		condition again, and continue waiting if it is not satisfied.
+	::
+
+## Have CondVar check your condition for you: call wait or waitFor with a predicate, which takes care of the three steps above.
+::
+
+The following code illustrates these two approaches:
+
+code::
+// Version 1: `wait` without predicate -- while loop required
+(
+var c = CondVar();
+// Our resource here will be a "message", which is set to some string when available,
+// and set to empty when unavailable. This resource is initially unavailable.
+var message = "";
+
+// We add these delays to be clear about which thread runs first. Try changing these values to see what happens!
+var waitThreadDelay = 1;
+var signalThreadDelay = 2;
+
+// These two forked threads will share the condition variable, `c`, and the condition state, `message`.
+fork {
+	waitThreadDelay.wait;
+	// It might be tempting to use `if` here instead of `while`, because we can see below that we will always exit
+	// `wait` when the message is non-empty. But this would lead to bugs when multiple threads start to wait, so it's
+	// best to get in the habit of always using `while` when calling `wait` without a predicate.
+	while { message.isEmpty } {
+		"Resource is unavailable, waiting now".postln;
+		c.wait;
+	};
+	"Resource is available, message is: '%'\n".postf(message);
+};
+
+fork {
+	signalThreadDelay.wait;
+	message = "meow, world!";
+	c.signalOne;
+	"Resource was made available and waiting thread was signaled".postln;
+};
+)
+
+// Version 2: `wait` with predicate -- no while loop required
+(
+var c = CondVar(), message = "";
+var waitThreadDelay = 1;
+var signalThreadDelay = 2;
+
+fork {
+	waitThreadDelay.wait;
+	// As a predicate, we pass a function indicating the condition we want to become true.
+	// We will continue immediately if it already is true. In any case, after returning from this `wait` call, we are
+	// certain that the message is not empty. No other threads run in the time between that check and the return from
+	// `wait`, so the only way for the predicate to become false is if it depends on state external to the interpreter,
+	// like the system time.
+	c.wait({ message.notEmpty });
```

Comment:
I don't have a strong opinion about trailing closure syntax, so whatever you think is easiest for people to understand is fine with me.

I've reflowed the code and comments in the examples to 70 characters, but I'd like to note that if you want this to be standard for contributed patches, it isn't very helpful to bring it up in PRs without documenting that somewhere.

Reply:
for what it's worth, i also agree that 60 or 70 character lines look best, there just isn't a convention for this so i did what was easiest under my default vim configuration.

i got the impression from past discussions that many non-expert users don't find trailing closure syntax intuitive, which is why i avoided it here. but again, i'll do whatever you think is best. i have a slight preference for keeping it consistent -- i.e. if we change all the waits we should change waitFor, too.

Reply:
We have both `wait` and `waitFor`, and (as james notes) `c.waitFor(5) { … }` really falls apart (this is not the case with `if`, because we are so used to it). So I'd tend towards keeping both parentheses (`c.wait({ … })`) in this case. But no strong opinion ...

Reply:
Let me preface this by saying I'm OK to leave it alone, on grounds that `waitFor(timeOut) { predicate }` looks clumsy and that examples should be syntactically consistent.

`if(cond) { true_branch }` and `do(number_or_array) { body }` don't fall apart, I think, because the concepts of a conditional branch or loop body resemble control structures that are familiar from other languages. A predicate here isn't quite one of those.

At the same time, I'll also note that literally just four hours ago, Daniel Mayer [commented](https://scsynth.org/t/some-observations/3464/37) on the forum, "Coming from Lisp, the frequent mix of different kinds of brackets looked odd to me. That can be overcome, though, with indentation and dropping unnecessary brackets (e.g. `do { }` instead of `do({ })`)." It isn't the first time I've heard that remark. I would not be at all surprised if we start seeing examples in the wild of `cond.wait { ... }` (and I'll probably do it this way in my own code).

> I've reflowed the code and comments in the examples to 70 characters, but I'd like to note that if you want this to be standard for contributed patches, it isn't very helpful to bring it up in PRs without documenting that somewhere.

I'm just reacting to what I see:

![wrap](https://user-images.githubusercontent.com/318301/112738210-bc7ddc80-8f9b-11eb-84bf-e4596cd9f91a.png)

Yeah, I think we should discourage this in helpfile examples. (FWIW, I don't recall this ever coming up before -- there's no trend that I should have noticed and proposed a style guide change. That is, we're noticing it *now*.)

In any case, with a few search engine queries and poking at a couple of pages on the wiki, I can't even *find* the official style guide. So... not today.

Reply:
https://github.com/supercollider/supercollider/wiki/Code-style-guidelines

Reply:
> Let me preface this by saying I'm OK to leave it alone, on grounds that waitFor(timeOut) { predicate } looks clumsy and that examples should be syntactically consistent.

OK, in this case I will just leave it the way it is.